File: examples/python_demo/catalog/book.py
"""Books — each book has one author and lives in the catalog."""

from catalog.author import Author


class Book:
    def __init__(
        self,
        title: str,
        author: Author,
        price: float,
        stock: int = 0,
    ) -> None:
        self.title: str = title
        self.author: Author = author
        self.price: float = price
        self.stock: int = stock

    def is_available(self, requested: int = 1) -> bool:
        # <uml-activity name="book_availability" granularity="statement">
        in_stock = self.stock >= requested
        priced = self.price > 0
        result = in_stock and priced
        return result
        # </uml-activity>

    def reserve(self, quantity: int) -> None:
        self.stock -= quantity

File: examples/csharp_demo/Catalog/Book.cs
namespace Catalog;

public class Book
{
    public string Title { get; set; }
    public Author Author { get; set; }
    public decimal Price { get; set; }
    public int Stock { get; set; }

    public Book(string title, Author author, decimal price, int stock = 0)
    {
        Title = title;
        Author = author;
        Price = price;
        Stock = stock;
    }

    public bool IsAvailable(int requested = 1)
    {
        // <uml-activity name="book_availability" granularity="statement">
        bool inStock = Stock >= requested;
        bool priced = Price > 0;
        bool result = inStock && priced;
        return result;
        // </uml-activity>
    }

    public void Reserve(int quantity)
    {
        Stock -= quantity;
    }
}
